[FLINK-11008][State Backends, Checkpointing]SpeedUp upload state files using multithread

klion26 opened the pull request:
## What is the purpose of the change

[This patch](https://issues.apache.org/jira/browse/FLINK-11008) wants to speedup upload state files using multithread.

## Brief change log

Using multithread when uploading state files such as https://github.com/apache/flink/pull/6777 for restoring.

## Verifying this change

This change added tests and can be verified as follows:

- Add a ExceptionRethrow UT in `RocksDBStateDataTransferTest#testMultiThreadUploadThreadPoolExceptionRethrow`
- Add a correctly test UT in `RocksDBStateDataTransferTest#testMultiThreadUploadCorrectly`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**yes**)
  - If yes, how is the feature documented? (**docs**)

klion26 commented:
cc @azagrebin

gyfora commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -20,31 +20,44 @@
+public class RocksDbStateDataTransfer {
+	private static final int READ_BUFFER_SIZE = 16 * 1024;
 
 	static void transferAllStateDataToDirectory(
```
Could you please make this method static while you are at it? :) Just for consistency and so we can integrate this in other tooling

klion26 replied:
hi @gyfora ,thank you for the review.  For consistency do you mean the method and the class have the same access modifiers?   and could you please tell me where will the method be integrated in.

klion26 commented:
I squashed the two commits and retriggered the Travis test.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:
```diff
@@ -403,85 +409,6 @@ private void cleanupIncompleteSnapshot(@Nonnull List<StateObject> statesToDiscar
-								stateHandleID,
-								new PlaceholderStreamStateHandle());
-						} else {
-							sstFiles.put(stateHandleID, uploadLocalFileToCheckpointFs(filePath));
```
I think this logic of sst/placeholders/misc file separation belongs here.
`RocksDbStateDataTransfer` should be responsible only for the upload.

What if instead of calling `uploadLocalFileToCheckpointFs`, we fill `sstFilesToUpload` and `miscFilesToUpload` of `Map<StateHandleID, Path>` in this loop and then call `RocksDbStateDataTransfer.uploadStateFiles` for each map? This `uploadStateFiles` could return `Map<StateHandleID, StreamStateHandle>` which could be put into the final `sstFiles`/`miscFiles`.

I think `ConcurrentHashMap` is not needed then.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -61,6 +74,144 @@ static void transferAllStateDataToDirectory(
+			ExecutorService executorService = createExecutorService(uploadingThreadNum);
+
+			try {
+				List<Runnable> runnables = createUploadRunnables(
```
I think using `Callable<StreamStateHandle>` (path -> StreamStateHandle, as it was in `uploadLocalFileToCheckpointFs`) is simpler and stateless, comparing to using `Runnable` and implicit adding results into the concurrent map.

What if we implement `createUploadCallables` and use `CompletableFuture.callAsync` to create `Map<StateHandleID, CompletableFuture<StreamStateHandle>>`? Then the futures (map.values()) could be waited for and the map could be converted into the result `Map<StateHandleID, StreamStateHandle>`.

klion26 replied:
Hi @azagrebin, when I want to implement this by using `Callable`, I can not find `java.util.concurrent.CompletableFuture.callAsync` function(JDK 8), after searching in flink project, I just find `org.apache.flink.runtime.rpc.MainThreadExecutable#callAsync`,  Did I miss something?

In other words, If I use `Callable` instead of `Runnable`, should I implement a 
 `ThrowingCallable` just as `ThrowingRunnable`.

If I use `Runnable` to implement this, how about create a `CompletableFuture<StreamStateHandle>` instance and pass to the called function `uploadLocalFileToCheckpointFs(Path, CompletableFuture)`, in `uploadLocalFileToCheckpointFs` we complete the passed in `CompletableFuture<StreamStateHandle>`,  such as below

```
Map<StateHandleID, CompletableFuture<StreamStateHandle>> futures = new HashMap();
// futures will update in createUploadRunnables
List<Runnable> runnables = createUploadRunnables(sstFilePaths, miscFilePaths, futures);
for (Runnable runnable : runnables) {
    CompletableFuture.runAsync(runnable, executorService);
}
FutureUtils.waitForAll(futures.values()).get();

....

private static StreamStateHandle uploadLocalFileToCheckpointFs(Path path, CompletableFuture future) {
   try {
       ...
       future.complete(...)
   } finally {
       ....
   }
}
```

azagrebin replied:
Hi @klion26 , I think this could work:
```
futures .put(stateHandleID , 
  CompletableFuture.supplyAsync(CheckedSupplier.unchecked(
    () -> uploadLocalFileToCheckpointFs(path)), executorService))
```
`CompletableFuture.supplyAsync` seems to do already internally what you suggest.
Also I would run parallel upload separately for  `sstFilePaths` and `miscFilePaths`, one after another, so that `RocksDbStateDataTransfer` does not need to know about file structure.

klion26 replied:
@azagrebin thank you for your reply, I'll update the code according to your suggestions.

klion26 commented:
@azagrebin thank you for your reviewing, today I have some other urgent things to do in my company, I'll update the code in the next two days.

klion26 commented:
@azagrebin I've updated the code according to your suggestions, please review again when you are free.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateDataTransferTest.java:
```diff
@@ -138,6 +148,157 @@ public void testMultiThreadRestoreCorrectly() throws Exception {
+		String localFolder,
+		int sstFileCount,
+		int fileStateSizeThreshold,
+		Map<StateHandleID, Path> stateFilePaths) throws IOException {
```
also here, I would rather return `stateFilePaths`.

klion26 commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateDataTransferTest.java:
```diff
@@ -138,6 +148,157 @@ public void testMultiThreadRestoreCorrectly() throws Exception {
+		Map<StateHandleID, StreamStateHandle> sstFiles = new HashMap<>(sstFileCount);
+		Map<StateHandleID, StreamStateHandle> miscFiles = new HashMap<>(miscFileCount);
+
+		RocksDbStateDataTransfer.uploadFilesToCheckpointFs(
```
@azagrebin In the current implementation, we upload the misc and sst files separately, so I want to test that both the misc and sst files are uploaded successfully.

azagrebin replied:
Indeed, it happens separately but in higher level code, not in `RocksDbStateDataTransfer.uploadFilesToCheckpointFs`. The two calls of `RocksDbStateDataTransfer.uploadFilesToCheckpointFs` in the test do logically the same thing and do not depend on file names or content.

klion26 replied:
@azagrebin got it, will change it.

klion26 commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -61,6 +69,88 @@ static void transferAllStateDataToDirectory(
+		int numberOfSnapshottingThreads,
+		CheckpointStreamFactory checkpointStreamFactory,
+		CloseableRegistry closeableRegistry,
+		Map<StateHandleID, StreamStateHandle> hanldes) throws Exception {
```
@azagrebin The implementation here returns `handles` implicitly because I thought the caller of the function `uploadFilesToCheckpointFs` may know the size of `handles`, and could init the map by `new HashMap(size)`. 
I agree that the map's size here would not be too large because we'are under incremental mode, I'll change to the explicitly mode. Anyway, I'll add a java doc for these two public function.

klion26 commented:
@azagrebin Thank you for your review, I'll rebase master and address the comments in the next commit.

klion26 commented:
@azagrebin I've rebased master and addressed all the comments. I can't fast-forward push the changed code to my remote branch, so I squashed the four commits and force pushed, sorry for the inconvenience.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java:
```diff
@@ -410,74 +412,51 @@ private void uploadSstFiles(
+					if (existsAlready) {
+						// we introduce a placeholder state handle, that is replaced with the
+						// original from the shared state registry (created from a previous checkpoint)
+						sstFiles.put(
```
let's keep short enough expressions as one line, also `final boolean existsAlready...`

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateDataTransferTest.java:
```diff
@@ -138,6 +149,133 @@ public void testMultiThreadRestoreCorrectly() throws Exception {
+	 */
+	@Test
+	public void testMultiThreadUploadCorrectly() throws Exception {
+
```
could you remove this new line?

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateDataTransferTest.java:
```diff
@@ -138,6 +149,133 @@ public void testMultiThreadRestoreCorrectly() throws Exception {
+
+		Map<StateHandleID, StreamStateHandle> sstFiles = new HashMap<>(sstFileCount);
+
+		sstFiles.putAll(RocksDbStateDataTransfer.uploadFilesToCheckpointFs(
```
I think it can be directly assigned to 
`Map<StateHandleID, StreamStateHandle> sstFiles = uploadFilesToCheckpointFs(..)`

klion26 commented:
@azagrebin Thank you for your quick response, and all the patience comments, I'vs addressed the comments. Thank you again.

StefanRRichter commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -61,6 +80,116 @@ static void transferAllStateDataToDirectory(
+
+		Map<StateHandleID, StreamStateHandle> handles = new HashMap<>();
+
+		ExecutorService executorService = createExecutorService(numberOfSnapshottingThreads);
```
This line makes me wonder if it is a good idea that this is just a static utils class instead of a proper class. One advantage of executor services is that they don't need to recreate threads all the time. I agree that this method is not called in any hot loops, so the overhead factor might be not dramatic, but still I wonder why not to make this a proper class that creates a threadpool once and only shuts it down at the end of the object lifecycle. Any reason why you chose this way?

klion26 replied:
@StefanRRichter I chose this way to align the download function `downloadDataForAllStateHandles`. 
I thought `uploadFilesToCheckpointFs` function may receive both the `sstFiles` and `miscFiles` as parameter, so we can reuse the executorService, but in that way, makes the function `uploadFilesToCheckpointFs` a little harder to reuse.

azagrebin replied:
@klion26 
The upload and download part are quite independent. The base class could contain the executor, number of threads, closable registry and close method (to shutdown the executor instead of shutdowning it every time). The  upload and download classes can extend it and share the executor. The downloader could be then registered with the closable registry. For upload, we can of course shutdown it immediately after restore is done. I agree that it would make sense to keep threads up and for code reuse as well.

StefanRRichter commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java:
```diff
@@ -218,7 +218,10 @@
+	private final int numberOfRestoringThreads;
+
+	/** Thread number used to upload to DFS when snapshot. */
+	private final int numberOfSnapshottingThreads;
```
Do you think it makes sense that user can configure both numbers independently? It might give more flexibility, but also makes it harder to configure or maybe a user can forget to set one of the two values. Did you experience scenarios where it made sense to have not just one number of io threads?

klion26 replied:
@StefanRRichter In my opinion, there may be a scenario that we use multithread to restore from the savepoint but do not need to use multithread to snapshot when incremental checkpoint.

In my company, we only enabled multithread restoring. we use only RocksDBIncremental Checkpoint to snapshot and restore from the latest completed checkpoint(we do not use savepoint to restore the job and do not use HeapStateBackend in the product).

StefanRRichter commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java:
```diff
@@ -61,6 +80,116 @@ static void transferAllStateDataToDirectory(
+	 * @param files The files will be uploaded to checkpoint filesystem.
+	 * @param numberOfSnapshottingThreads The number of threads used to upload the files.
+	 * @param checkpointStreamFactory The checkpoint streamFactory used to create outputstream.
+	 * @param closeableRegistry
```
Missing javadoc.

klion26 commented:
@StefanRRichter @azagrebin thank you for your review, I've addressed the comments.

The tests failed because the doc is outdated, I'll push another commit soon.

I changed the config name and a class name(typo before), I think the changed config&class name have not been released ever, so it would not affect users. Please let me know if this behavior should be reverted.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/AsyncSnapshotCallable.java:
```diff
@@ -97,6 +97,10 @@ protected void cancel() {
 	}
 
+	protected CloseableRegistry getSnapshotCloseableRegistry() {
+		return snapshotCloseableRegistry;
```
could we just make `snapshotCloseableRegistry` protected and use it directly in `RocksIncrementalSnapshotStrategy`? I think we do not need `(un)registerCloseableForCancellation` methods either afterwards and replace calls to them with direct calls to `snapshotCloseableRegistry.(un)registerCloseable`.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java:
```diff
@@ -517,6 +516,10 @@ public void restore(Collection<KeyedStateHandle> restoreState) throws Exception
 
 		LOG.info("Initializing RocksDB keyed state backend.");
 
+		if (LOG.isDebugEnabled()) {
```
this log seems to be a duplicate of the following one in try/if/else.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBOptions.java:
```diff
@@ -47,10 +47,10 @@
 
 	/**
-	 * The number of threads used to download files from DFS in RocksDBStateBackend.
+	 * The number of threads used to transfer(download and upload) files in RocksDBStateBackend.
```
space in `transfer(download` before `(`, also below in similar places

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -701,17 +702,22 @@ public RocksDBNativeMetricOptions getMemoryWatcherOptions() {
+	/**
+	 * Sets the number of threads used to transfer files while snapshotting/restoring.
+	 *
+	 * @param numberOfTransferingThreads The number of threads used to download files from DFS while restoring.
```
`used to download files from DFS while restoring` -> `to transfer files while snapshotting/restoring`

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateDownloader.java:
```diff
@@ -36,43 +36,47 @@
 
-		downloadDataForAllStateHandles(sstFiles, dest, restoringThreadNum, closeableRegistry);
-		downloadDataForAllStateHandles(miscFiles, dest, restoringThreadNum, closeableRegistry);
+		downloadDataForAllStateHandles(sstFiles, dest, closeableRegistry);
```
could `closeableRegistry` be shared the same way as `executorService` in `RocksDBStateDataTransfer`? Then we do not need to pass to all class methods.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateUploader.java:
```diff
@@ -0,0 +1,157 @@
+/**
+ * Help class for uploading RocksDB state.
+ */
+public class RocksDBStateUploader extends RocksDBStateDownloader {
```
`extends RocksDBStateDataTransfer` not `RocksDBStateDownloader`?

klion26 replied:
My bad, I chose the wrong one when using auto-complete.

klion26 commented:
Seems like there is a relevant test failed, I'm looking into it.

klion26 commented:
@azagrebin thank you for your review, I've updated the code based on your suggestions. Please have another look when you are free.

azagrebin commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateUploader.java:
```diff
@@ -0,0 +1,153 @@
+ * Help class for uploading RocksDB state files.
+ */
+public class RocksDBStateUploader extends RocksDBStateDataTransfer {
+	private final int readBufferSize = 16 * 1024;
```
can be static
